### PR TITLE
Added UI for managing flat rate packages

### PR DESCRIPTION
**Pull request by robobot3000**
Adds #670 

Should be tested with `add/centralized-flat-rates` checked out on the server-side.

To test:
- Check that you can see the predefined packages in the Packaging Manager:
screen shot 2016-11-07 at 14 03 49
- Check that disabling the packages hides the corresponding services from the shipping instance settings:
screen shot 2016-11-07 at 14 05 39
- Check that no flat rates are returned to the customers when flat rate packages are disabled in the packaging manager

### CODE REVIEW

**Comment by allendav** [comment]
@kellychoffman @robobot3000 - to be consistent with our other UX, how about checkboxes on the left instead?

screen shot 2016-11-07 at 11 44 28 am

**Comment by allendav** [comment]
The number of options for USPS flat rate packaging is overwhelming. Since all domestic flat rate boxes can be used for international flat rate (with one exception), let's not have separate checkboxes for domestic and international for a given size.

The one exception is the APO/FPO/DPO flat rate box. APO/FPO/DPO addresses are for US armed forces overseas, but they are treated like domestic addresses.  Supporting this flat rate box will probably require further code changes (i.e. to detect APO/FPO/DPO addresses), so defer this particular box to a separate issue.

**Comment by allendav** [comment]
Let's also not bother with Gift Card Flat Rate Envelopes at this time.  Please exclude them from the UI.

**Comment by allendav** [comment]
Let's not bother with the DVD Box nor the Large Video Box at this time. Please exclude them from the UI.

**Comment by allendav** [comment]
Instead of a single group of "USPS Flat Rate Packages" lets break this up a bit so that merchants have an easier time finding what they need.  I propose the following grouping (cc'ing @kellychoffman  for her insights)

Let's also take care to name these exactly as shown below (both groups and packages) since these are the exact names on the USPS website and the names the merchant is most likely to understand

(Note that there are two Window Flat Rate Envelopes - I've included the dimensions in parentheses, but the name should not include the dimensions since we display them separately)

screen shot 2016-11-07 at 12 35 58 pm

**Comment by allendav** [comment]
Neither "Post Card" nor "Letter" are flat rate packaging. Please do not include in the UI.

screen shot 2016-11-07 at 12 40 59 pm

**Comment by allendav** [comment]
I could not find this flat rate packaging on the USPS web site at all. Maybe it doesn't exist / we entered it in error?

screen shot 2016-11-07 at 12 42 18 pm

**Comment by allendav** [comment]
@robobot3000 - if combining domestic and international for each package is going to substantially blow up the work for this PR, then it is OK for us to have this PR NOT do that - in that case we should 1) write a new issue to combine domestic and international for each package and 2) have six groups (three domestic groups and three international groups).  The group names should be something like

USPS Priority Mail Flat Rate Envelopes
USPS Priority Mail Flat Rate Boxes
USPS Priority Mail Express Flat Rate
USPS Priority Mail International Flat Rate Envelopes
USPS Priority Mail International Flat Rate Boxes
USPS Priority Mail Express International Flat Rate

**Comment by robobot3000** [comment]
Most of the requested changes had to be added on the server side in Automattic/woocommerce-connect-server#619. I disabled the services that you requested to hide from the UI (gift cards, videos etc). I also added groups and merged the domestic and international flat rates. See the server PR for more info

File: client/packages/views/index.js [hunk 1]
```diff
@@ -34,7 +84,12 @@ const Packages = ( props ) => {
 			<CompactCard className="settings-group-card">
 				<FormSectionHeading className="settings-group-header">{ __( 'Packaging' ) }</FormSectionHeading>
```
Comment (DanReyLop) [inline client/packages/views/index.js]: If the other section is called `Prefecined packages`, I would call this one something more consistent, like `Custom packages`.

File: client/packages/views/packages-list.js [hunk 2]
```diff
@@ -16,18 +18,28 @@ const noPackages = () => {
...
+			return null;
+		}
+
+		const allPackageIds = packages.map( ( def ) => def.id );
+		const selectedAll = 0 === _.difference( allPackageIds, selected ).length;
+		return <FormLegend className="package-actions"><FormCheckbox checked={ selectedAll } onChange={ () => toggleAll( serviceId, groupId ) }/></FormLegend>;
+	};
 
 	const renderPackageListItem = ( pckg, idx ) => {
+		const isSelected = selected && -1 !== selected.indexOf( pckg.id );
```
Comment (DanReyLop) [inline client/packages/views/packages-list.js]: `Array.prorotype.includes()` is much cleaner than the `-1` thing. Let's make good use of babel :) [Docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes)

File: client/settings/views/services/group.js [hunk 3]
```diff
@@ -63,14 +64,18 @@ const ShippingServiceGroup = ( props ) => {
 				</label>
 			</div>
 
-			{ services.map( ( service, idx ) => (
-				<ShippingServiceEntry
+			{ services.map( ( service, idx ) => {
+				if ( service.predefined_package && -1 === predefinedPackages.indexOf( service.predefined_package ) ) {
```
Comment (DanReyLop) [inline client/settings/views/services/group.js]: Same comment here about `Array.prototype.includes`

**Comment by allendav** [comment]
@robobot3000 - this looks much better!  But I get the following error when checking a box:

screen shot 2016-11-08 at 12 51 42 pm

**Comment by robobot3000** [comment]
This happens only if the packages were never defined (the option doesn't exist in the database). It's now fixed

**Comment by allendav** [comment]
Thank you @robobot3000 - it tests nicely now

File: client/packages/views/index.js [hunk 4]
```diff
@@ -13,8 +15,56 @@ import * as PackagesActions from '../state/actions';
...
+		const elements = [];
+
+		_.forEach( props.form.predefinedSchema, ( servicePackages, serviceId ) => {
+			const serviceSelected = props.form.packages.predefined[ serviceId ] || [];
+
+			_.forEach( servicePackages, ( predefGroup, groupId ) => {
+				const groupPackages = predefGroup.definitions;
+				const groupPackageIds = groupPackages.map( ( def ) => def.id );
+				const diffLen = _.difference( groupPackageIds, serviceSelected ).length;
+				const summary = 0 === diffLen ? __( 'All packages selected' ) : sprintf( __( '%d packages selected' ), groupPackageIds.length - diffLen );
```
Comment (allendav) [inline client/packages/views/index.js]: When only 1 package is selected, this displays "1 packages selected" - let's handle that case better ("1 package selected")

**Comment by allendav** [comment]
I think we can 🚢 this ... @DanReyLop do you concur?

see also https://github.com/Automattic/woocommerce-connect-server/pull/619

**Comment by DanReyLop** [comment]
Yes, OK to 🚢 for me

**Pull request merged**